### ISSUE TITLE
ViewEncapsulation.None styles bleed into ViewEncapsulation.Native components when siblings

### ISSUE THREAD

**Issue by gnomeontherun** [angular/angular#5059]
When there are sibling components, one with ViewEncapsulation.Native and the other with ViewEncapsulation.None, the styles that any component stylings are being appended to the shadow root of the one component set to Native. 

Plunker: http://plnkr.co/edit/huN8xe0S7xH5CUw4B2P6?p=preview

<img width="491" alt="screen shot 2015-11-01 at 9 58 03 pm" src="https://cloud.githubusercontent.com/assets/79500/10874083/afa88d42-80e3-11e5-8f8e-ad09cd8414df.png">

A component set to Native should not adopt the rest of the CSS rules from non-Native components. In the screenshot of the rendered DOM, the last three style elements are appended to the end of the shadow root, but they are actually declared in the other two components which are set to None and Emulated modes.

If all three components are set to Native, then the rules are applied only to the local component, but when they are mixed the rules bleed over into Native (but don't bleed out of Native).

**Comment by tbosch** [comment]
Right now Angular is doing this so a user can mix and match components with
different ViewEncapsulations.
If we didn't do this, using a component with ViewEncapsulation.None/
Emulated inside of a component with ViewEncapsulation.Native would not work
correctly regarding styling, as shadow roots blocks off all styles from the
document head.

**Comment by gnomeontherun** [comment]
That is interesting, but as a counterpoint wouldn't that also potentially cause issues with the Native mode? If I'm using a third party component that is None and it uses the same CSS selector, it would get injected into my component. I certainly didn't expect sibling style rules to get injected into the shadow root of the Native mode component.

I can at least see a case for more granular control over if this occurs or not, perhaps with a StrictNative mode.

**Comment by mlaval** [comment]
No longer the case, see http://plnkr.co/edit/XDIJ1YOz4W3uI9t75Jg4?p=preview

**Comment by mlaval** [comment]
Still the case in fact, cf plnkr above

**Comment by dudewad** [comment]
@tbosch Re: your comment above:
`as shadow roots blocks off all styles from the document head.`
 Does this apply to `ViewEncapsulation.Emulated` as well? I am seeing that is the case when using `ViewEncapsulation.None` inside elements that have the default encapsulation (i.e. not specifying encapsulation). The styles do not take.
I'm not sure if the issue is related to this one or not.

**Comment by splincode** [comment]
@gnomeontherun Is the issue relevant?

**Comment by gnomeontherun** [comment]
Depends on what you mean by relevant. I found this a while back with my investigations of encapsulation modes and believe this is at least not documented still today. I would have called this a bug, but there are reasons it could make sense, but basically anytime you mix modes you're probably playing with 🔥 

I've not run into this issue with real world scenarios, but that is because I don't mix modes.

**Comment by splincode** [comment]
@gnomeontherun Please close the problem, or rediscover it according to the template. At the moment I do not see this issue relevant..

**Comment by splincode** [comment]
@gnomeontherun Should we reopen the issue?

Since your demo does not work:
![image](https://user-images.githubusercontent.com/12021443/43358426-2809bae8-929a-11e8-94b6-e9592a1387dc.png)

**Comment by gnomeontherun** [comment]
Looks like the plunkr broke sometime in the past few years, but I think this is still a problem. It is at least not what I would expect, as Native should shield you from external styles, where this explicitly injects them from sibling components.

**Comment by collinstevens** [comment]
I'm experiencing a similar issue, asked here https://stackoverflow.com/questions/53129175/viewencapsulation-emulated-styles-are-copied-into-shadow-root-by-a-viewencapsul/53135513#53135513. I've also got a simple reproduction here https://github.com/collinstevens/angular-encapsulation.

**Comment by petebacondarwin** [comment]
@tbosch wrote:

> Right now Angular is doing this so a user can mix and match components with
different ViewEncapsulations.
> If we didn't do this, using a component with ViewEncapsulation.None/
Emulated inside of a component with ViewEncapsulation.Native would not work
correctly regarding styling, as shadow roots blocks off all styles from the
document head.

As far as I can tell we have a thing called a `SharedStylesHost`, and each ShadowHost is attached to this, as well as the top level host (i.e. the top level `<document>`). Whenever we create a new ShadowHost, all the styles from the top level host are copied into the new ShadowHost, so that the styles for `None` and `Emulated` components are available.

This doesn't seem right to me.

Instead, I believe that we should create a new `SharedStylesHost` for each ShadowHost that does not inherit all the styles from the top level. Then, if a component with `ViewEncapsulation` set to `Emulated` or `None` is used within this ShadowHost, the component should add its styles only to the nearest ancestor that has a `SharedStylesHost`.  I think this would solve the problem of `None` mode styles being available within the ShadowHost but also prevent unwanted styles from appearing in the ShadowDOM from outside.

**Comment by petebacondarwin** [comment]
Actually this is the same issue as in #7887 where @tbosch actually described the same situation as above. But didn't offer a solution. I'll close that one as although it has some good discussion, there is nothing more than is here already.

**Comment by JBusch** [comment]
This is still the case with ViewEncapsulation.Shadowdom which is very bad imho. 

First, it creates huge overhead inserting all that styles, but even more concerning, you can not depend on the correct rendering of your components because there are styles injected which you explicitly not want in it.

That is the whole point of shadow dom, right?

**Comment by petebacondarwin** [comment]
Right. I believe the solution is what I wrote here https://github.com/angular/angular/issues/5059#issuecomment-766167662

**Comment by JBusch** [comment]
@petebacondarwin would it be possible to currently use your approch? like overwriting the SharedStylesHost ?

**Comment by petebacondarwin** [comment]
I am not sure. The `SharedStylesHost` is a global service that is injected into the DOM renderer. I think we would need to change things so that this could be a tree of hosts... Probably somewhere in the `DomRendererFactory2` class...

**Comment by JBusch** [comment]
Too bad, thank you :)
